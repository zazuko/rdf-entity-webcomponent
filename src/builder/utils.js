import { ns } from '../namespaces.js'
import rdf from '../rdf-ext.js'

function subjects (cf) {
  return [...rdf.termSet(quads(cf).map(quad => quad.subject))]
}

function predicates (cf) {
  return [...rdf.termSet(quads(cf).map(quad => quad.predicate))]
}

function quads (cf) {
  return [...cf.dataset.match(cf.term, null, null, null, null)]
}
function splitIfVocab (iri) {
  const candidates = Array.from(Object.entries(ns)).filter(([_, value]) => {
    return iri.startsWith(value().value)
  })
  if (candidates.length) {
    candidates.sort(([, iri1], [, iri2]) => iri2.length - iri1.length)
    const found = candidates[0]

    return {
      string: iri.replace(new RegExp(`^${found[1]().value}`), ''),
      vocab: found[0]
    }
  }

  const lastSegment = iri.split('/').pop()
  return {
    string: lastSegment
  }
}

export { subjects, predicates, quads, splitIfVocab }
