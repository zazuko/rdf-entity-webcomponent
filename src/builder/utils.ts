import { ns } from '../namespaces'
import {rdf} from '../rdf-ext'
import {GraphPointer} from "clownface";
import {VocabValue} from "../types";

function predicates (cf:GraphPointer) {
  return [...rdf.termSet(quads(cf).map(quad => quad.predicate))]
}

function quads (cf:GraphPointer) {
  return [...cf.dataset.match(cf.term, null, null, null)]
}

function splitIfVocab (iri:string):VocabValue {
  const candidates = Array.from(Object.entries(ns)).filter(([_, value]) => {
    return iri.startsWith(value().value)
  })

  if (candidates.length) {
    candidates.sort(([, iri1], [, iri2]) => iri2.length - iri1.length)
    const found = candidates[0]

    return {
      value: iri.replace(new RegExp(`^${found[1]().value}`), ''),
      vocab: found[0]
    }
  }
  const lastSegment = iri.endsWith('/')
    ? iri.slice(0, -1).split('/').pop()
    : iri.split('/').pop()
  return {
    value: lastSegment
  }
}

export { predicates, quads, splitIfVocab }
