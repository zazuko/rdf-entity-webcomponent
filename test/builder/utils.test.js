import { expect } from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { describe, it } from 'mocha'
import { splitIfVocab } from '../../src/builder/utils.js'

expect.extend({ toMatchSnapshot })

const entities = [
  'https://example.org/people/',
  'http://www.w3.org/2001/XMLSchema#string',
  'https://example.org/people']

describe('utils', () => {
  for (const named of entities) {
    it(`splitIfVocab ${named}`, function () {
      const result = splitIfVocab(named)
      expect(result).toMatchSnapshot(this)
    })
  }
})
