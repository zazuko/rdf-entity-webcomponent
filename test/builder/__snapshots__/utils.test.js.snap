// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils splitIfVocab https://example.org/people 1`] = `
Object {
  "value": "people",
}
`;

exports[`utils splitIfVocab https://example.org/people/ 1`] = `
Object {
  "value": "people",
}
`;
